src/libsync/propagatedownloadzsync.cpp

```diff
@@ -0,0 +1,291 @@
+}
+
+bool GETFileZsyncJob::finished()
+{
+    if (reply()->bytesAvailable()) {
+        return false;
+    }
+
+    if (_current < _nrange - 1) {
+        _current++;
+        startCurrentRange(_zbyterange.get()[2*_current],_zbyterange.get()[(2*_current)+1]);
+        return false;
+    }
+
+    if (!_hasEmittedFinishedSignal) {
+        emit finishedSignal();
+    }
+
+    _hasEmittedFinishedSignal = true;
+
+    return true; // discard
+}
+
+void GETFileZsyncJob::seedFinished(void* zs)
+{
+    _zs = zsync_unique_ptr<struct zsync_state> ((struct zsync_state*)zs, [](struct zsync_state* zs) {
+        zsync_complete(zs);
+        zsync_end(zs);
+    });
+    if (!_zs) {
+        _errorString = tr("Unable to parse zsync.");
+        _errorStatus = SyncFileItem::NormalError;
+        qCDebug(lcZsyncGet) << _errorString;
+        emit finishedSignal();
+        return;
+    }
+
+    {   /* And print how far we've progressed towards the target file */
+        long long done, total;
+
+        zsync_progress(_zs.get(), &done, &total);
+        qCInfo(lcZsyncGet).nospace() << "Done reading: "
+            << _localpath << " " << fixed << qSetRealNumberPrecision(1) <<  (100.0f * done) / total
+            << "% of target seeded.";
+
+    }
+
+    /* Get a set of byte ranges that we need to complete the target */
+    _zbyterange = zsync_unique_ptr<off_t> (zsync_needed_byte_ranges(_zs.get(), &_nrange, 0), [](off_t* zbr) {
+        free(zbr);
+    });
+    if (!_zbyterange) {
+        _errorString = tr("Failed to get zsync byte ranges.");
+        _errorStatus = SyncFileItem::NormalError;
+        qCDebug(lcZsyncGet) << _errorString;
+        emit finishedSignal();
+        return;
+    }
+
+    qCDebug(lcZsyncGet) << "Number of ranges:" << _nrange;
+
+    /* If we have no ranges then we have equal files and we are done */
+    if (_nrange == 0) {
+        _propagator->reportFileTotal(*_item, 0);
+        _errorStatus = SyncFileItem::FileIgnored;
+        emit finishedSignal();
+        return;
+    }
+
+    _zr = zsync_unique_ptr<struct zsync_receiver> (zsync_begin_receive(_zs.get(), 0), [](struct zsync_receiver* zr) {
+       zsync_end_receive(zr);
+    });
+    if(!_zr) {
+        _errorString = tr("Failed to initialize zsync receive structure.");
+        _errorStatus = SyncFileItem::NormalError;
+        qCDebug(lcZsyncGet) << _errorString;
+        emit finishedSignal();
+        return;
+    }
+
+    quint64 totalBytes = 0;
+    for (int i=0; i<_nrange; i++) {
+        totalBytes += _zbyterange.get()[(2*i)+1] - _zbyterange.get()[(2*i)];
+    }
+    qCDebug(lcZsyncGet) << "Total bytes:" << totalBytes;
+    _propagator->reportFileTotal(*_item, totalBytes);
+
+    /* start getting bytes for first zsync byte range */
+    startCurrentRange(_zbyterange.get()[0],_zbyterange.get()[1]);
+}
+
+void GETFileZsyncJob::seedFailed(QString errorString)
+{
+    _errorString = errorString;
+    _errorStatus = SyncFileItem::NormalError;
```

Comment:
None yet, definitely need some support to understand best approach here, both of which you mentioned, but which and when and how, I'm not sure.

Reply:
Also worth noting that `GETFileZsyncJob::seedFailed` is not expected to be called, it would usually mean a corrupt .zsync file on remote. So I would think deleting the remote .zsync is the best/only option.

Reply:
Done.